File: frontend/src/lib/digest.ts
import { getDaysUntilDate, parseDateOnly } from './reminderDisplay'
import { toAttentionReminder, type AttentionReminder } from './reminderSchedule'
import type { Reminder, ReminderAlert, ReminderType } from '../types/reminder'

export interface DigestOptions {
  lookaheadDays: number
}

export interface DigestSmartGroup {
  type: Extract<ReminderType, 'birthday' | 'renewal' | 'maintenance'>
  label: string
  count: number
}

export interface DailyDigest {
  needsAttention: AttentionReminder[]
  dueToday: Reminder[]
  comingUp: Reminder[]
  smartGroups: DigestSmartGroup[]
  totals: {
    needsAttention: number
    dueToday: number
    comingUp: number
  }
}

const sectionLimit = 5

const smartGroupLabels: Record<DigestSmartGroup['type'], string> = {
  birthday: 'Birthdays coming up',
  renewal: 'Renewals coming up',
  maintenance: 'Maintenance coming up',
}

export function buildDailyDigest(reminders: Reminder[], alerts: ReminderAlert[], options: DigestOptions): DailyDigest {
  const usedReminderIds = new Set<string>()
  const activeReminders = reminders.filter((reminder) => !reminder.completed)
  const needsAttentionAll = alerts.map(toAttentionReminder)
  needsAttentionAll.forEach((item) => usedReminderIds.add(item.reminder.id))

  const dueTodayAll = activeReminders
    .filter((reminder) => !usedReminderIds.has(reminder.id) && getDaysUntilDate(reminder.due_date) === 0)
    .sort(sortByDueDateThenTitle)
  dueTodayAll.forEach((reminder) => usedReminderIds.add(reminder.id))

  const comingUpAll = activeReminders
    .filter((reminder) => {
      if (usedReminderIds.has(reminder.id)) {
        return false
      }

      const daysUntil = getDaysUntilDate(reminder.due_date)
      return daysUntil > 0 && daysUntil <= options.lookaheadDays
    })
    .sort(sortByDueDateThenTitle)

  const smartGroups = getSmartGroups(comingUpAll)

  return {
    needsAttention: needsAttentionAll.slice(0, sectionLimit),
    dueToday: dueTodayAll.slice(0, sectionLimit),
    comingUp: comingUpAll.slice(0, sectionLimit),
    smartGroups,
    totals: {
      needsAttention: needsAttentionAll.length,
      dueToday: dueTodayAll.length,
      comingUp: comingUpAll.length,
    },
  }
}

export function getDigestSummaryText(digest: DailyDigest) {
  const counts = [
    formatDigestCount(digest.totals.needsAttention, 'needs attention'),
    formatDigestCount(digest.totals.dueToday, 'due today'),
    formatDigestCount(digest.totals.comingUp, 'coming up'),
  ]

  return counts.join(' • ')
}

export function hasDigestItems(digest: DailyDigest) {
  return digest.totals.needsAttention > 0 || digest.totals.dueToday > 0 || digest.totals.comingUp > 0
}

function getSmartGroups(reminders: Reminder[]): DigestSmartGroup[] {
  const counts: Record<DigestSmartGroup['type'], number> = {
    birthday: 0,
    renewal: 0,
    maintenance: 0,
  }

  reminders.forEach((reminder) => {
    if (reminder.reminder_type === 'birthday' || reminder.reminder_type === 'renewal' || reminder.reminder_type === 'maintenance') {
      counts[reminder.reminder_type] += 1
    }
  })

  return (Object.keys(counts) as DigestSmartGroup['type'][])
    .filter((type) => counts[type] > 0)
    .map((type) => ({
      type,
      label: smartGroupLabels[type],
      count: counts[type],
    }))
}

function formatDigestCount(count: number, label: string) {
  return `${count} ${label}`
}

function sortByDueDateThenTitle(left: Reminder, right: Reminder) {
  const dueDifference = parseDateOnly(left.due_date).getTime() - parseDateOnly(right.due_date).getTime()

  if (dueDifference !== 0) {
    return dueDifference
  }

  return left.title.localeCompare(right.title)
}

File: backend/app/digest.py
from dataclasses import dataclass
from datetime import date, datetime, timedelta

from app.alerts import get_alert_eligibility, sort_alerts
from app.models import Reminder


@dataclass(frozen=True)
class DailyDigestSummary:
    needs_attention: int
    due_today: int
    coming_up: int

    @property
    def has_items(self) -> bool:
        return self.needs_attention > 0 or self.due_today > 0 or self.coming_up > 0

    def to_push_body(self) -> str:
        return (
            f"{self.needs_attention} needs attention "
            f"\u2022 {self.due_today} due today "
            f"\u2022 {self.coming_up} coming up"
        )


def build_daily_digest_summary(
    reminders: list[Reminder],
    *,
    lookahead_days: int,
    now: datetime,
    current_day: date | None = None,
) -> DailyDigestSummary:
    resolved_current_day = current_day or now.date()
    used_reminder_ids: set[str] = set()
    active_reminders = [reminder for reminder in reminders if not reminder.completed]

    alert_reminders = []
    for reminder in reminders:
        eligibility = get_alert_eligibility(reminder, now=now, current_day=resolved_current_day)
        if eligibility is not None:
            alert_reminders.append((reminder, eligibility))
            used_reminder_ids.add(reminder.id)

    sorted_alerts = sort_alerts(alert_reminders)

    due_today = [
        reminder
        for reminder in active_reminders
        if reminder.id not in used_reminder_ids and reminder.due_date == resolved_current_day
    ]
    for reminder in due_today:
        used_reminder_ids.add(reminder.id)

    lookahead_end = resolved_current_day + timedelta(days=lookahead_days)
    coming_up = [
        reminder
        for reminder in active_reminders
        if reminder.id not in used_reminder_ids and resolved_current_day < reminder.due_date <= lookahead_end
    ]

    return DailyDigestSummary(
        needs_attention=len(sorted_alerts),
        due_today=len(due_today),
        coming_up=len(coming_up),
    )
